### ISSUE TITLE
Governance API questions and improvements

### ISSUE THREAD

**Issue by jbibla** [luniehq/lunie#4812]
1. can depositers and voters be identity objects like proposer (name, address, image if possible)?
2. can proposer be an object instead of a string?
3. timeline and links don't need to be inside of detailedVotes
4. proposal descriptions should always include the content of the proposal. eg: `/polkadot/proposals/democracy-1` has been replaced with our "description" but the content of the parameter change is missing. let's always leave the actual proposal description in the description field and add our content as a `note` (https://kusama.polkassembly.io/referendum/81)
5. timeline should include all the steps of the governance process (including future steps - array of 3 or 4 objects) and one of the objects should be `current: true` to indicate where the proposal is in the timeline
6. can beneficiary be an object instead of a string?
7. what is the `validator` field for?
8. can we have actual IDs like polkassembly (https://kusama.polkassembly.io/referendum/81 appears as `democracy-1` for us)?
9. `percentageDepositsNeeded` is null - and why is it percentage as opposed to total number of tokens?
10. `/polkadot/proposals/democracy-1` has no date / time information at all anywhere ... is this expected?
11. `polkadot/proposals/referendum-2` all the votes (163) have the same `voter` - is `voter` supposed to be address? ideally this would be an identity object

**Comment by Bitcoinera** [comment]
> can depositers and voters be identity objects like proposer (name, address, image if possible)?
> can proposer be an object instead of a string?

You mean like creating a new API Object for proposer? This doesn't exist. Proposer is always a String, like you said

**Comment by jbibla** [comment]
> You mean like creating a new API Object for proposer?

yeah - if you look at the designs (😉 ) you would see what the requirements were intended to be ... the FE needs a name, address, and image always. for votes and deposits we need extra info (amount with denom and time):

<img width="369" alt="Screen Shot 2020-09-02 at 10 21 47 AM" src="https://user-images.githubusercontent.com/6021933/91995685-1cfab300-ed06-11ea-96ca-b6dccc3a1ef7.png">

<img width="1105" alt="Screen Shot 2020-09-02 at 10 22 53 AM" src="https://user-images.githubusercontent.com/6021933/91995821-43b8e980-ed06-11ea-961c-d17cb8a083c9.png">

**Comment by Bitcoinera** [comment]
OK, I see. Getting the names and so on for proposers and voters is quite easy. The other thing with the deposits times I need to investigate a bit, but yeah, surely possible

**Comment by Bitcoinera** [comment]
So what about a type called `NetworkAccount`?

It would be:

```
{
   name
   address
   image
}
```

All these, voter, depositer and proposer would have this type

**Comment by jbibla** [comment]
i imagined something even more generic like `identity` because we use accounts to mean something else (includes a key). 

it's a type of `identity` or `user`?

---

noticed another issue:

12. dates on cosmos are different than date formats for polkadot:
`2020-06-04T18:00:11.292428073Z`
`Sat, 22 Aug 2020 14:59:24 GMT`

the API should standardize everything so the FE can be dumb!

**Comment by jbibla** [comment]
13. can we make proposal timeline messages capitalized and remove the word "proposal"

```
Proposal created
Proposal deposit period ends
Proposal voting period starts
Proposal voting period ends
```

```
Created
Deposit Period Ends
Voting Period Started
Voting Period Ends
```

**Comment by jbibla** [comment]
14. timeline is null for council proposal `polkadot/proposals/council-14`

**Comment by jbibla** [comment]
15. i don't think council proposals are a proposal `type` - they are kind of something else. more like `source` or part of `proposer` - maybe `proposerGroup` and if it's `null` then that's fine.

**Comment by jbibla** [comment]
16. for this table (i'm calling it participant table) - is there a way to get only 5?

<img width="1105" alt="Screen Shot 2020-09-02 at 10 22 53 AM" src="https://user-images.githubusercontent.com/6021933/91995821-43b8e980-ed06-11ea-961c-d17cb8a083c9.png">

**Comment by faboweb** [comment]
> can we have actual IDs like polkassembly (https://kusama.polkassembly.io/referendum/81 appears as democracy-1 for us)?

We would need to index all proposals to give them ids. How are they getting deterministic IDs? What is the problem with the current system?

**Comment by Bitcoinera** [comment]
> 1. timeline is null for council proposal `polkadot/proposals/council-14`

Well, this is expected. I couldn't find a way to get even the start date for a council proposal. But I can research further, of course

**Comment by Bitcoinera** [comment]
> 1. i don't think council proposals are a proposal `type` - they are kind of something else. more like `source` or part of `proposer` - maybe `proposerGroup` and if it's `null` then that's fine.

:eyes: :question:  you lost me there

**Comment by faboweb** [comment]
council proposals are a proposal to add someone as a council member or a proposal made by the council?

**Comment by Bitcoinera** [comment]
> council proposals are a proposal to add someone as a council member or a proposal made by the council?

The process to enter the council is different. I think it depends on stake and votes from regular accounts, but not 100%. I can ask around

**Comment by Bitcoinera** [comment]
> proposal descriptions should always include the content of the proposal. eg: /polkadot/proposals/democracy-1 has been replaced with our "description" but the content of the parameter change is missing. let's always leave the actual proposal description in the description field and add our content as a note (https://kusama.polkassembly.io/referendum/81)

I don't get what you mean here? Mmm, what I see is that this one is actually a referendum, so something must be failing there :thinking:

**Comment by jbibla** [comment]
@faboweb 

> We would need to index all proposals to give them ids. How are they getting deterministic IDs? What is the problem with the current system?

it's fine how it is (for milestone completion) but it's going to cause confusion for people who look at other explorers and sites

**Comment by Bitcoinera** [comment]
> /polkadot/proposals/democracy-1 has no date / time information at all anywhere ... is this expected?
Yes, the problem with democracy proposals is when they have no image object within. Then it is quite impossible to get the metadata. When it looks like this, then I have found virtually no way to do it:

```
{
"balance": 1000000000000,
"seconds": [
"15MrYzj6KsJAom1V3eTH4Wgsz86FLVE65dF7SEeNkPYSsSs7",
"167m6sRoKcR3EvyYRu5sGstczgEYNnC2qS61AruU3yRxJa7y",
"14BNXJMDTGHjXirVpiHwXY8Yv7SgPSCL5eB6jrhvoaBq9DRm",
"162L864FSfPg2YhKuqMu6CKyosWWjmeCfqnqZaukbt6aQfaZ",
"147zHFsxqfdjHCPq1e71mV9F94U9Qe7tPP6R5Cfb1tJ3XLer",
"12mjz14W7Dn4zZDQuqb4dUxmrHSEXU2pgkXpBb2urfERcKvn",
"13gppLNLGAkbgGHpnYYBLuQWJvHL5Ao2iibMerQFEg9RQzy9",
"12C6Kfs4hnwYkDvThKjyt7run9htQJ4GWcRTm7xSXqtnox6D"
],
"imageHash": "0x086650c6a3966e8179e5213fdb8bcc01b109d82d58c492c6f2c198f4237ec3cb",
"index": 1,
"proposer": "15MrYzj6KsJAom1V3eTH4Wgsz86FLVE65dF7SEeNkPYSsSs7"
},
```

With just the imageHash but without the creation block for the proposal it is quite hard to get any metadata. The only query that seems to help is `api.query.democracy.preimages()`, but this one I am not even sure how to use it. I get no response doing just `api.query.democracy.preimages(imageHash)` and I only see examples using `preimages.at( blockHash, imageHash)`. Getting the preimage I am not sure we can get the creation block either

**Comment by Bitcoinera** [comment]
> percentageDepositsNeeded is null - and why is it percentage as opposed to total number of tokens?

I will ask around. Not sure how many seconds a democracy proposal needs to get to become a refenrendum.

About the percentage, this is the field name in the epic, so this is why it is a percentage :shrug: We can change it to total number of tokens if you prefer

**Comment by Bitcoinera** [comment]
Starting to see this wrapping up
